pkg/tfshim/sdk-v2/upgrade_state.go

```diff
@@ -6,92 +6,148 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
 )
 
-func upgradeResourceState(ctx context.Context, p *schema.Provider, res *schema.Resource,
-	instanceState *terraform.InstanceState) (*terraform.InstanceState, error) {
+func upgradeResourceStateGRPC(
+	ctx context.Context, t string, res *schema.Resource,
+	state cty.Value, meta map[string]any,
+	server tfprotov5.ProviderServer,
+) (cty.Value, map[string]any, error) {
+	// TODO[pulumi/pulumi-terraform-bridge#1667]: This is not quite right but we need
+	// the old TF state to get it right.
+	jsonBytes, err := ctyjson.Marshal(state, state.Type())
+	if err != nil {
+		return cty.Value{}, nil, err
+	}
 
-	if instanceState == nil {
-		return nil, nil
+	version, _, err := extractSchemaVersion(meta)
+	if err != nil {
+		return cty.Value{}, nil, err
 	}
 
-	m := instanceState.Attributes
+	//nolint:lll
+	// https://github.com/opentofu/opentofu/blob/2ef3047ec6bb266e8d91c55519967212c1a0975d/internal/tofu/upgrade_resource_state.go#L52
+	if version > int64(res.SchemaVersion) {
+		return cty.Value{}, nil, fmt.Errorf(
+			"State version %d is greater than schema version %d for resource %s. "+
+				"Please upgrade the provider to work with this resource.",
+			version, res.SchemaVersion, t,
+		)
+	}
 
-	// Ensure that we have an ID in the attributes.
-	m["id"] = instanceState.ID
+	// Note upgrade is always called, even if the versions match
+	//nolint:lll
+	// https://github.com/opentofu/opentofu/blob/2ef3047ec6bb266e8d91c55519967212c1a0975d/internal/tofu/upgrade_resource_state.go#L72
 
-	version, hasVersion := 0, false
-	if versionValue, ok := instanceState.Meta["schema_version"]; ok {
-		versionString, ok := versionValue.(string)
-		if !ok {
-			return nil, fmt.Errorf("unexpected type %T for schema_version", versionValue)
-		}
-		v, err := strconv.ParseInt(versionString, 0, 32)
-		if err != nil {
-			return nil, err
-		}
-		version, hasVersion = int(v), true
+	resp, err := server.UpgradeResourceState(ctx, &tfprotov5.UpgradeResourceStateRequest{
+		TypeName: t,
+		RawState: &tfprotov5.RawState{JSON: jsonBytes},
+		Version:  version,
+	})
+	if err != nil {
+		return cty.Value{}, nil, err
 	}
 
-	// First, build a JSON state from the InstanceState.
-	json, version, err := schema.UpgradeFlatmapState(ctx, version, m, res, p.Meta())
+	newState, err := msgpack.Unmarshal(resp.UpgradedState.MsgPack, res.CoreConfigSchema().ImpliedType())
 	if err != nil {
-		return nil, err
+		return cty.Value{}, nil, err
+	}
+
+	newMeta := make(map[string]interface{}, len(meta))
+	// copy old meta into new meta
+	for k, v := range meta {
+		newMeta[k] = v
 	}
+	newMeta["schema_version"] = strconv.Itoa(res.SchemaVersion)
 
-	// Next, migrate the JSON state up to the current version.
-	json, err = schema.UpgradeJSONState(ctx, version, json, res, p.Meta())
+	return newState, newMeta, nil
+}
+
+func extractSchemaVersion(meta map[string]any) (int64, bool, error) {
```

Comment:
This needed to be extracted out since we use it in multiple places:
- We use it in `upgradeResourceStateGRPC` to send the current schema version to the provider server.
- We use it in `upgradeResourceState` to find the correct type to hydrate the instanceState.Attributes into.